### PR TITLE
Header madness

### PR DESCRIPTION
**Pull request by profxj**
Another `try/except` to handle header crashes.

Am guessing this was similar in spirit to what @jhennawi 
proposed awhile back.

Added a test.

And also confirmed that if the user-supplied PypeIt file
trumps all (didn't find a great test for that).

### CODE REVIEW

**Comment by profxj** [comment]
Note:  The user still has to turn on 

```
[rdx]
spectrograph = keck_lris_red
ignore_bad_headers = True
```

for `run_pypeit` to work successfully on data with bad headers.

**Comment by profxj** [comment]
=============== 288 passed, 4503 warnings in 1248.14s (0:20:48) ================

**Review by kbwestfall — APPROVED** [review]
Looks good.  My only comment is that we don't have a good doc on what to do if you've got bad headers.  Outside of the pypeit parameters doc, we mention the `ignore_bad_headers` keyword in the `metadata.rst` doc, which is really buried in the html hierarchy (it's only accessibly via the development doc).  Maybe we should add something to the `pypeit_file.rst` doc describing use of this parameter and edits to the pypeit file to deal with bad header values?

**Comment by profxj** [comment]
--- PYPEIT DEVELOPMENT SUITE FAILED 1/124 TESTS (Masters ignored) ---
Failed tests:
    keck_deimos/900ZD_M_6000 pypeit (without masters)
Skipped tests:
Testing Started at 2021-08-30T21:44:16.943756
Testing Completed at 2021-08-31T04:43:06.830231
Total Time: 6:58:49.886475


The failure is unrelated to this PR..

**Comment by profxj** [comment]
Have added a `Known Failure Modes` doc.

check it out, @kbwestfall 

https://github.com/pypeit/PypeIt/blob/header_madness/doc/known_failure_modes.rst

File: doc/index.rst [hunk 1]
```diff
@@ -136,6 +154,7 @@ What this version provides
   * :doc:`2d_combine`
   * :doc:`coadd1d`
   * :doc:`coadd2d`
+  * :doc:`known_failure_modes`
```
Comment (kbwestfall) [inline doc/index.rst]: I think it would be better if we put this in one of the toc trees below so that it shows up in the left sidebar.

Comment (profxj) [reply doc/index.rst]: ah, that was my intent.  

Have done so now.

**Comment by profxj** [comment]
=============== 288 passed, 4503 warnings in 1262.99s (0:21:02) ================

DevSuite running..

File: pypeit/spectrographs/spectrograph.py [hunk 2]
```diff
@@ -790,26 +793,37 @@ def get_meta_value(self, inp, meta_key, required=False, ignore_bad_header=False,
...
-            try:
+                # Grab from the header, if we can
                 value = headarr[self.meta[meta_key]['ext']][self.meta[meta_key]['card']]
-            except (KeyError, TypeError):
-                value = None
+        except (KeyError, TypeError) as e:
+            if ignore_bad_header or (not required):
+                msgs.warn("Bad Header, but we'll try to continue on..") 
+            else:
+                embed(header='817 of spectro')
```
Comment (kbwestfall) [inline pypeit/spectrographs/spectrograph.py]: Remove this `embed`?

**Comment by profxj** [comment]
--- PYPEIT DEVELOPMENT SUITE PASSED 128/128 TESTS (Masters ignored) ---
Testing Started at 2021-09-01T16:58:09.951007
Testing Completed at 2021-09-01T22:12:49.903387
Total Time: 5:14:39.952380

File: pypeit/spectrographs/spectrograph.py (hunk 2 above)
Comment (profxj) [reply pypeit/spectrographs/spectrograph.py]: gone

File: doc/known_failure_modes.rst [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+to add the following to your :ref:`pypeit_file`::
+
+    [rdx]
+    ignore_bad_headers = True
+
+Note that PypeIt always uses the entries in your data block
+(often slurped from the headers) as the *true* values, i.e.
+you can modify those by hand to over-come aspects of your
+corrupt header.
+
```
Comment (jhennawi) [inline doc/known_failure_modes.rst:44]: Thanks, I'm very happy with this.

File: pypeit/spectrographs/spectrograph.py [hunk 4]
```diff
@@ -790,26 +793,36 @@ def get_meta_value(self, inp, meta_key, required=False, ignore_bad_header=False,
...
 
         # Is this not derivable?  If so, use the default
         #   or search for it as a compound method
         value = None
-        if self.meta[meta_key]['card'] is None:
-            if 'default' in self.meta[meta_key].keys():
-                value = self.meta[meta_key]['default']
-            elif 'compound' in self.meta[meta_key].keys():
-                value = self.compound_meta(headarr, meta_key)
+        try:
```
Comment (jhennawi) [inline pypeit/spectrographs/spectrograph.py:810]: Try statements with 20 lines of code below them are both not good coding practice and very difficult to debug. I am as guilty of this as anyone, but my suggestion is that we try to purge as many try except blocks from this routine as possible, and if we do feel we have to use them, then only try: except: a single line of code (as best coding practice). I recently spent hours trying to find a trivial bug (going line by line) related to this routine that was obscured by  all the try: except: madness.

File: pypeit/spectrographs/spectrograph.py [hunk 5]
```diff
@@ -790,26 +793,36 @@ def get_meta_value(self, inp, meta_key, required=False, ignore_bad_header=False,
...
-            try:
+                # Grab from the header, if we can
                 value = headarr[self.meta[meta_key]['ext']][self.meta[meta_key]['card']]
-            except (KeyError, TypeError):
-                value = None
+        except (KeyError, TypeError) as e:
+            if ignore_bad_header or (not required):
+                msgs.warn("Bad Header, but we'll try to continue on..") 
+            else:
+                raise e
```
Comment (jhennawi) [inline pypeit/spectrographs/spectrograph.py:825]: Please consider whether we can remove the try: except blocks below.

File: pypeit/tests/test_pypeitsetup.py [hunk 6]
```diff
@@ -168,3 +151,21 @@ def test_run_setup():
     os.remove(data_path('shane_kast_blue.sorted'))
 
 
+@dev_suite_required
```
Comment (jhennawi) [inline pypeit/tests/test_pypeitsetup.py:154]: Thanks for adding this test. If I understand this is now a unit test that tries to run various operations on files known to have bad headers? Does this also test pypeit_obslog?

**Review by jhennawi — CHANGES_REQUESTED** [review]
Minor suggestion about get_meta_value. The control flow for this routine is opaque, and I fear more try: except statements will only make it more opaque.

File: pypeit/tests/test_pypeitsetup.py (hunk 6 above)
Comment (profxj) [reply pypeit/tests/test_pypeitsetup.py:154]: yes, that is right

**Comment by profxj** [comment]
yup, the `try/except` aren't fun but it is a symptom
of the flexibility we need to have for some header
cards of some instruments to be sometimes required
and sometimes not (e.g. RA/DEC of VLT/Xshooter)

File: pypeit/spectrographs/spectrograph.py (hunk 4 above)
Comment (jhennawi) [reply pypeit/spectrographs/spectrograph.py:810]: This is not resolved. Please try and except the individual lines that can fail due to the following errors and not the entire code block.

**Comment by jhennawi** [comment]
@profxj I agree with you about the flexibility that is needed, but disagree that it is necessary to try/except a long block of code in order to achieve that flexibility. That makes debugging very difficult, which is an experience I just had trying to debug exactly this routine. For that reason, I am asking that your try/excepts test individual lines that can fail. If that means adding 10 more try:excepts, then so be it. In my opinion, get_meta_value needs to be completely refactored. In the absence of that, you can try:except the individual lines that could fail.

File: pypeit/spectrographs/spectrograph.py (hunk 5 above)
Comment (jhennawi) [reply pypeit/spectrographs/spectrograph.py:825]: Which lines of code in the block above can fail due to KeyErrror or TypeError? Looks to me like it is only line 820 and possibly also line 815. If compound_meta is parsing headers you should just be passing ignore_bad_header to that routine as well, and then there can be a try:except at exactly the line that can fail. That makes the code easier to read and also easier to debug.

**Comment by profxj** [comment]
I'm ok closing this and leaving for a future day.

But I'm not refactoring.

**Comment by profxj** [comment]
i.e. @jhennawi -- either accept this and I'll 
raise an Issue to refactor someday or close
it and we'll have #1249 still open.

Would be unfortunate to lose the docs, but someone can
dig those out eventually.

**Review by jhennawi — APPROVED** [review]

**Comment by jhennawi** [comment]
Okay that sounds good, please raise the issue.

**Comment by profxj** [comment]
merging

**Pull request merged**